src/Roslyn.Suppressions/Roslyn.Suppressions/SettingsFile/SuppressedIssuesFileWatcher.cs

```diff
@@ -0,0 +1,71 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2022 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using System.IO.Abstractions;
+
+namespace SonarLint.VisualStudio.Roslyn.Suppressions.SettingsFile
+{
+    internal interface ISuppressedIssuesCache
+    {
+        void Invalidate(string settingsKey);
+    }
+
+    internal sealed class SuppressedIssuesFileWatcher : IDisposable
+    {
+        private readonly IFileSystemWatcher fileSystemWatcher;
+        private readonly ISuppressedIssuesCache suppressedIssuesCache;
+
+        public SuppressedIssuesFileWatcher(ISuppressedIssuesCache suppressedIssuesCache)
+            : this(suppressedIssuesCache, new FileSystem())
+        {
+        }
+
+        internal SuppressedIssuesFileWatcher(ISuppressedIssuesCache suppressedIssuesCache, IFileSystem fileSystem)
+        {
+            this.suppressedIssuesCache = suppressedIssuesCache;
+
+            fileSystem.Directory.CreateDirectory(RoslynSettingsFileInfo.Directory);
```

Comment:
From manual testing, `FileSystemWatcher.FromPath` will throw argument exception if the directory does not exist. In practice it _shouldn't_ happen, but maybe it's better to safeguard just in case.

Reply:
Yep, better to be defensive.
There are a couple of scenarios in which the directory might not exist:
* first use; the external process might have started running before SLVS has had a chance to create the directory for the first time;
* Windows clears up the temp folder

Reply:
Hm, that's a good question. If we try-catch the exception, it will leave VS' external process in an inconsistent state where our stuff are partially initialized (i.e. suppressions code is there but the file watcher was never created, so the cache would never update). Perhaps it would be better to fail, throw, and let the caller handle it, so that next time we would try to re-create it.